### PR TITLE
[#20] Allows to define a static ip to use after booting

### PR DESCRIPTION
**Pull request by DT3264**
[#20] Allows to define a static ip to use after booting

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/equetzal/huronOS-build-tools/pull/107).
* #173
* __->__ #107

### CODE REVIEW

File: base-system/usrroot/usr/sbin/ipman
Comment (equetzal) [inline base-system/usrroot/usr/sbin/ipman:40]: We should either ask for the network mask or infer it from the given IP address, but the network mask can be different than this.

File: base-system/usrroot/usr/sbin/ipman
Comment (equetzal) [inline base-system/usrroot/usr/sbin/ipman:43]: Not sure if there's a way to actually check directly with connman.

File: base-system/usrroot/usr/sbin/ipman
Comment (equetzal) [inline base-system/usrroot/usr/sbin/ipman:47]: Even if ICMP protocol for pings is widely open, there could be a case in which http, tcp port 80 is open but ICMP protocol is blocked. Not sure if we can rely on this. Is there some way to check this with connman directly?
Also `ip addr` could be a better indicator of link up and running.

File: base-system/usrroot/usr/lib/systemd/system/ipman.service
Comment (equetzal) [inline base-system/usrroot/usr/lib/systemd/system/ipman.service:8]: What about `After=network.target` so that we're sure that we're running after network target. 
Also we could also make a dependency with Connman service directly.

File: base-system/usrroot/usr/sbin/ipman
Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:40]: Got it, added support for ip masks in the server-sync.conf

File: base-system/usrroot/usr/lib/systemd/system/ipman.service
Comment (DT3264) [reply base-system/usrroot/usr/lib/systemd/system/ipman.service:8]: You're right, Im kinda new on the services, so this flew over my head

File: base-system/usrroot/usr/sbin/ipman
Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:43]: Connman seems to have a status command, but at lest in huronOS it stays at `State = Ready` if an IP is configured on the interface (even if is not a valid ip / there's no response from the gateway for any request), yet it does not transition to the `State = Online` even if the connection is successful and requests to the outside have a response.

File: base-system/usrroot/usr/sbin/ipman
Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:47]: > Even if ICMP protocol for pings is widely open, there could be a case in which http, tcp port 80 is open but ICMP protocol is blocked. Not sure if we can rely on this.

You're right, I wasn't aware on that posibility,
Wdyt on the following solution?

 ```
if [[ $(ping -q -c 3 -W 5 "$IP_SERVER" >/dev/null) || $(curl -s "$DIRECTIVES_FILE_URL" /dev/null) ]]; then`
```

I think that it might work, because if the ping fails (lets say ICMP is blocked but the network is available) then the curl should work, otherwise we can say that we have no internet connection.


>Is there some way to check this with connman directly?
Also ip addr could be a better indicator of link up and running.

On the connman side, as mentioned on another review
>>Connman seems to have a status command, but at lest in huronOS it stays at State = Ready if an IP is configured on the interface (even if is not a valid ip / there's no response from the gateway for any request), yet it does not transition to the State = Online even if the connection is successful and requests to the outside have a response.

Regarding ip addr, although that could be a good indicator if the link is up and running, and after some playing with the command, seems like it lacks the capability of aknowledging if the network has connection to the outside.

Comment (equetzal) [reply base-system/usrroot/usr/sbin/ipman]: This option sounds good, I'll be traveling for now but as soon as I can test on hardware I'll check how smooth does this option works.

File: base-system/usrroot/usr/sbin/ipman
Comment (equetzal) [inline base-system/usrroot/usr/sbin/ipman:45]: What happens with Wifi interfaces? I don't think the config will ever work because we don't have any network passwords config.
Maybe we can shrink the scope to only ethernet¿, we know all ethernet interfaces can only start the `e`

**Review by equetzal — COMMENTED** [review]
Generally looks good, seems to work and it adds a super important feature on the system. Great work on this!

Some thoughts:
- After reading better the code and trying it, probably we need to re-work the naming to something more explicit and standard. We have `$IP_ADDRESS` and `$DIRECTIVES_SERVER_IP`, probably we need to change both to something explicit as `$LOCAL_IP_ADDRESS` and `$DIRECTIVES_SERVER_IP_ADDRESS`. Also it could be `$INSTANCE_IP_ADDRESS`, `$LOCAL_IP_ADDRESS`, `$CLIENT_IP_ADDRESS`, `$HOST_IP_ADDRESS`. Not sure which one to choose, but surely we need to change it so that it doesn't get out of control as we add more and more vars.
- What happens if by some reason we loss connectivity during the runtime? How can someone come and fix the IP issue? just by re-running the service? Can hsync re-launch the service if needed? Also, shouldn't hsync depend on this service too?

Overall LGTM, but let's handle those comments.

**Review by equetzal — CHANGES_REQUESTED** [review]
Sorry, forgot to set it to request changes for the bad-configured comment.

**Comment by DT3264** [comment]
> What happens if by some reason we loss connectivity during the runtime? How can someone come and fix the IP issue? just by re-running the service? Can hsync re-launch the service if needed? Also, shouldn't hsync depend on this service too?

Good question, In theory reruning the script might do the work.

>Can hsync re-launch the service if needed? 

Regarding hsync re-launch, I'm not sure if that's possible, do you know a way to archieve that? 🤔 

>Also, shouldn't hsync depend on this service too?

You're right, I'll add the dependency on this service.

File: software-modules/base/03-budgie/budgie.sh
Comment (DT3264) [reply software-modules/base/03-budgie/budgie.sh:143]: Okay, added the chmod on prepare.sh

File: base-system/tools/installer/install.sh
Comment (DT3264) [reply base-system/tools/installer/install.sh:106]: Sounds good, I'll open an issue for that one

File: base-system/tools/installer/install.sh
Comment (DT3264) [reply base-system/tools/installer/install.sh:116]: > we tell that those vars can only be used in conjunction with the IP_ADDRESS.

You mean as a warning only?

File: base-system/tools/installer/install.sh
Comment (DT3264) [reply base-system/tools/installer/install.sh:1]: Done

File: base-system/usrroot/usr/sbin/ipman
Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:45]: Done by grepping to lines starting with e

File: base-system/tools/installer/install.sh
Comment (equetzal) [reply base-system/tools/installer/install.sh]: > > we tell that those vars can only be used in conjunction with the IP_ADDRESS.
> 
> You mean as a warning only?, or is it okay to still exit if mask / gateway are set but the instance address isn't?

If one is present, all 3 params must be present to work; so in that case is ok to exit. But in this conditional, if one of them is empty (possibly the 3), you exit, which is not the intension as the static IP should not be a installer requirement.

File: base-system/tools/installer/install.sh
Comment (equetzal) [inline base-system/tools/installer/install.sh:106]: I would prefer that instead of `--instance-ip-address|-ia` you only set the large one even if internally it's possible the short one. 
Let's do another PR for setting a proper help page.

File: base-system/tools/installer/install.sh
Comment (equetzal) [inline base-system/tools/installer/install.sh:119]: What about a message like:
```terminal
Configuration error on
[--instance-ip-address $IP --instance-ip-mask $NETMASK --instance-ip-gateway $GATEWAY]

IP Address, Gateway and Network Mask are required when setting static IP address for the huronOS instance.

Run `./install.sh --help` for more information.
```

File: base-system/tools/installer/install.sh
Comment (equetzal) [inline base-system/tools/installer/install.sh:116]: You can simplify the condition to:
if (any of this vars is non-empty) and (any of this vars is empty); then ...

**Review by equetzal — CHANGES_REQUESTED** [review]
After correcting the problems with the PR, I've tested and been able to set the static IP correctly; I've also been able to fallback to DHCP when using a wrong configuration or if the config doesn't match the network it correctly uses DHCP.

Please address the problems of the review and then I'll be able to AC.

File: base-system/tools/installer/install.sh
Comment (DT3264) [reply base-system/tools/installer/install.sh:119]: Sounds good, I like that mesage.

File: base-system/tools/installer/install.sh
Comment (DT3264) [reply base-system/tools/installer/install.sh:106]: Okay, on my way to make a PR on top of this one with help

File: base-system/tools/installer/install.sh
Comment (DT3264) [reply base-system/tools/installer/install.sh:116]: Thanks for the suggestion, added.

File: base-system/usrroot/usr/sbin/ipman
Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:19]: My bad, updating those too.

File: base-system/usrroot/usr/sbin/ipman
Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:21]: Updated too

File: base-system/usrroot/usr/sbin/ipman
Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:20]: Updated too

File: base-system/tools/installer/install.sh
Comment (DT3264) [reply base-system/tools/installer/install.sh]: Improved the execution by following your suggestion from later on.

File: base-system/prepare.sh
Comment (DT3264) [reply base-system/prepare.sh:62]: Okay, added the chmod.
And I agree, although I'm thinking that once a PR is in review it cannot go to a "do not review this PR yet, I'm working on it rn" state, or can it?
It would be quite useful so I save you time on mid-baking PRs

File: docs/usage/how-to-install.md
Comment (DT3264) [reply docs/usage/how-to-install.md:90]: Okay, adding another PR on top, though I'll make a quick change in the meantime to have the stack ready, but I'll add those doc comments asap.

File: base-system/prepare.sh
Comment (equetzal) [reply base-system/prepare.sh:62]: > Okay, added the chmod. And I agree, although I'm thinking that once a PR is in review it cannot go to a "do not review this PR yet, I'm working on it rn" state, or can it? It would be quite useful so I save you time on mid-baking PRs

I know, Github UI is confusing haha, I had the same problem. You can set it to re-set it to draft again in the reviewers section:
Captura de pantalla 2023-08-04 a la(s) 8 58 51 a m
This is usually done when you'll be doing hard changes to the PR. When their are only light you usually send changes and re-ask for review when needed. It might be a good idea to later setup a bot for specifically asking for reviews and ignore the Github notifications about updates; most companies have something similar to this but I've not explored that for now. 
For now just feel free to setup to draft your PRs until you feel they are ready to be reviewed again! :D

File: docs/usage/how-to-install.md
Comment (equetzal) [reply docs/usage/how-to-install.md:90]: > Okay, but before, where would be a good path to put said doc? And I guess it should be references after the `fill the prompts` regarding the static IP

Probably under the `usage/` directory, you can name it `usage/using-static-ip.md`; you can follow the instructions on this paper https://github.com/equetzal/huronOS-build-tools/pull/155, It still WIP but I'll finish it as soon as possible.

Comment (equetzal) [reply docs/usage/how-to-install.md:90]: Also, feel free to send your docs PR after merging this one, but don't forgot about it. 
Just waiting for you to send the corrected changes and I'll approve this one. 
Nice work!

Comment (DT3264) [reply docs/usage/how-to-install.md:90]: I already sent the corrections, take a look and let me know if I forgot something 😝

Comment (equetzal) [reply docs/usage/how-to-install.md:90]: > I already sent the corrections, take a look and let me know if I forgot something 😝

Not sure if the current latest commit is updated. Can you please submit it again on sapling? I don't think I'm watching the updated version.

Comment (DT3264) [reply docs/usage/how-to-install.md:90]: I re-submitted the PR, let's see if it went well

**Review by equetzal — APPROVED** [review]
LGTM!!
Congrats on this!!

File: base-system/usrroot/usr/sbin/ipman
Comment (equetzal) [inline base-system/usrroot/usr/sbin/ipman:71]: What happens if I didn't set the directives file on purpose? 
The script will just discard the connection even if the config is ok?

Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:71]: That's an interesting question, I guess that if you only leave out the directives file, it will still keep that interface as configured if and only if the ip of the directives file server is set, thus I'm thinking:
what would be a reliable default url / ip we could fetch / ping if either or neither of the variables are set?

Comment (DT3264) [reply base-system/usrroot/usr/sbin/ipman:71]: @equetzal 
The review might have been because from the comment above this one.

**Review by equetzal — APPROVED** [review]
I've tested this again, and everything seems to be working fine!
The change to ping the gateway was really good improvement!
Feel free to merge

**Pull request merged**
